### ISSUE TITLE
RuleForEach not validated immediately

### ISSUE THREAD

**Issue by Ben555555** [Blazored/FluentValidation#161]
When I used a Model with a collection, it seems like the collection items are not validated immediately (when I enter and leave a form element), unlike the properties on the Model itself. Only when I submit the form it will be validated:

Validators

```
public class TransportModelValidator : AbstractValidator<TransportModel>
{
        public TransportModelValidator(TransportStepModelValidator transportStepModelValidator)
        {
              // validated immediately
             RuleFor(model => model.Description)
                .MaximumLength(255);

               RuleForEach(model => model.TransportSteps)
                .SetValidator(transportStepModelValidator);
        }
}

public class TransportStepModelValidator : AbstractValidator<TransportStepModel>
{
        public TransportStepModelValidator()
        {
            // not validated immediately
            RuleFor(model => model.ExpectedDate)
                .NotEmpty();
        }
}
```

Blazor

```
   <FluentValidationValidator DisableAssemblyScanning="@true" />
 <MudTextField                    
                                Margin="Margin.Dense"
                                Variant="Variant.Outlined"
                                Label="@Frontend.Disposition_Transports_TransportStep_Description"
                                @bind-Value="Model.Description"
                                For="() => Model.Description" />

  @foreach(var ts in Model.TransportSteps)
    {
        <MudDatePicker
            Margin="Margin.Dense"
            @bind-Date="ts.ExpectedDate"
            Label="@Frontend.Common_Date"
            Editable="true"
            For="() => ts.ExpectedDate"
            Variant="Variant.Outlined"
            Mask="InputMasks.DateMask"
            AutoClose="true">
        </MudDatePicker>
        <ValidationMessage For="@(() => ts.ExpectedDate)" />
    }
```

Is there any way to control this behaviour?

**Comment by Sharaf-Mansour** [comment]
@Ben555555 downgrade to 2.0.1 or use the pre-release 
this issue was addressed in 2.0.3

**Comment by Sharaf-Mansour** [comment]
@Ben555555 can you provide a tiny repo with the issue? 
So, I can take a better look.

**Comment by GamingCoder** [comment]
I experimented around with this, because I had the same use case. **You have to have AbstractValidators for the outer and inner type inside the assembly AND set the validator for the inner type in the outer validator.** When a field is changed the validator is determined based on the type of the model. For this to work currently both validators are required. During form submission only the validator for the outer model is validated and therefore has to include the inner validator.

This also means that any validation rules specified for the inner type in the validator for the outer type are only respected during form submission, but I at least can live with this.

**Comment by sven5** [comment]
I have exactly the same issue.

My validator looks like

```
 public class ProductVariantValidator : AbstractValidator<ProductVariantViewModel>
 {
     public ProductVariantValidator()
     {
         RuleForEach(p => p.Properties).ChildRules(c => c.RuleFor(p => p.Value)
             .NotEmpty().WithMessage(p => $"Please set a value for {p.Name}."));
     }
 }
```

**Comment by Sharaf-Mansour** [comment]
@sven5 give me more details. dotnet version and library version.
 Is it blazor wasm or blazor server?

**Comment by sven5** [comment]
Hi @Sharaf-Mansour 
.NET 7 and Blazor Server with newest library version 2.1.0

**Comment by Sharaf-Mansour** [comment]
@sven5 I think I know the issue.
It is the `ChildRules`
Mind if I make a quick demo. Because I think there is other way Is to create a `Dedicated validator` for the child 
```cs
public class CertificateValidator : AbstractValidator<Certificate>
{
    public CertificateValidator()
    {
        RuleFor(x => x.Date).SetValidator(new DateValidator());
        RuleFor(x => x.Data).SetValidator(new DataValidator());
    }
}
```
I will try to recreate your scenario on my end just to make sure it is exactly what I expected to be.

**Comment by sven5** [comment]
> Because I think there is an other way. By creating a `Dedicated validator` for the child

Yes, perhaps that would be working. However, I wanted to avoid this extra step.

**Comment by Sharaf-Mansour** [comment]
@sven5 Well it is working.
You will need to go one more level deeper with the `validator` and make a `dedicated validator` to get a live detection of the **validation**, Otherwise it will only trigger on submit.
I think that is because how `assembly scanners` work in general. It match and map the exact validation for the child not triggering its parent validation because the type will mismatch.
Feel free to take a look at [this folder of my repo](https://github.com/Sharaf-Mansour/CVBuilder/tree/main/ModelValidators) because that is how I got them to work. 

Take a look at `DataValidator.cs`, `CertificateValidator.cs` and also `CVValidator.cs`
You will notice that all of those are linked. but if we remove the link.
The validation will work `onchange` but not `onsubmit` and if we keep the link but make the same thing you did **Having them all on the same level** the `onchange` wont work but the `onsubmit` will work. so if you want the best of both worlds. do both. it worked for me. hopefully will work for you. 

### I know that this is weird but, the way how scanners work. it just match and map. nothing fancy.

**Comment by sven5** [comment]
@Sharaf-Mansour Thanks for clarification.
However, it's still not working in my particular case.

**Comment by Sharaf-Mansour** [comment]
@sven5 can you make a small minimal repo with the classes you want to validate that has the same issue.
Or show me the final code snippet you used. 
So I can provide more help.

**Comment by sven5** [comment]
@Sharaf-Mansour Thanks for your support. I got it working now with the new validator for the child class as you suggested. 😄
